src/peer_channels.rs

```diff
@@ -0,0 +1,611 @@
+//! 3. The first step of this routine is to introduce yourself with a regular message containing the `IrohPublicKey`.
+//!    This message contains the users relay-server and public key.
+//!    Direct IP address are not included as this information can be persisted by email providers.
+//! 4. After the announcement, the sending peer joins the gossip swarm with an empty list of peer IDs (as they don't know anyone yet).
+//! 5. Upon receiving an announcement message, other peers store the sender's [NodeAddr] in the database
+//!    (scoped per WebXDC app instance/message-id). The other peers can then join the gossip with `joinRealtimeChannel().setListener()`
+//!    and `joinRealtimeChannel().send()` just like the other peers.
+
+use anyhow::{anyhow, Context as _, Result};
+use iroh_gossip::net::{Gossip, JoinTopicFut, GOSSIP_ALPN};
+use iroh_gossip::proto::{Event as IrohEvent, TopicId};
+use iroh_net::magic_endpoint::accept_conn;
+use iroh_net::relay::{RelayMap, RelayUrl};
+use iroh_net::{key::SecretKey, relay::RelayMode, MagicEndpoint};
+use iroh_net::{NodeAddr, NodeId};
+use rand::{thread_rng, Rng};
+use url::Url;
+
+use crate::chat::send_msg;
+use crate::config::Config;
+use crate::context::Context;
+use crate::message::{Message, MsgId, Viewtype};
+use crate::mimeparser::SystemMessage;
+use crate::EventType;
+
+impl Context {
+    /// Create magic endpoint and gossip for the context.
+    pub(crate) async fn inite_peer_channels(&self) -> Result<()> {
+        let secret_key: SecretKey = self.get_or_generate_iroh_keypair().await?;
+        info!(self, "Iroh secret key: {}", secret_key.to_string());
+
+        let mut ctx_gossip = self.gossip.lock().await;
+        if ctx_gossip.is_some() {
+            warn!(
+                self,
+                "Tried to create endpoint even though there already is one"
+            );
+            return Ok(());
+        }
+
+        let endpoint = Box::pin(
+            MagicEndpoint::builder()
+                .secret_key(secret_key)
+                .alpns(vec![GOSSIP_ALPN.to_vec()])
+                .relay_mode(
+                    /* self.metadata
+                    .read()
+                    .await
+                    .as_ref()
+                    .map(|conf| {
+                        let url = conf
+                            .iroh_relay
+                            .as_deref()
+                            .unwrap_or("https://iroh.testrun.org:4443");
+                        let url = RelayUrl::from(Url::parse(url)?);
+                        Ok::<_, url::ParseError>(RelayMode::Custom(RelayMap::from_url(url)))
+                    })
+                    .transpose()?
+                    // This should later be RelayMode::Disable as soon as chatmail servers have relay servers
+                    .unwrap_or(RelayMode::Default), */
+                    RelayMode::Custom(RelayMap::from_url(RelayUrl::from(
+                        Url::parse("https://iroh.testrun.org:4443").unwrap(),
```

Comment:
Better to make it the default value of a new config key so that it can be reconfigured via environment

Reply:
Gonna care about it when we are there. If its fine with you I'm resolving this for now..